dts: arm: st: Introduce pinctrl nodes and uart pinmux configuration

erwango opened the pull request:
Introduce pinctrl nodes and uart pinctrl configuration in stm32 socs

A preambule for this change is the rework of the pinmux driver in each series in order
to achieve a model where pin could be configured with the call to a single array
of {pin, pinconfig} couples. This rework should be tansparent in term of existing pinmux
features.
Then a pinctrl yaml description and pinctrl binding are introduced to be used with a pinctrl node.

Pinmux configurations are introduced in <soc>-pinctrl.dtsi files. This is done only for u(s)art for now.
All uart pin configurations in existing c files are reproduced here. Once we'll be able to 
generate code thanks to dts files, clean up should be done to remove duplicated information.

ydamigos commented on dts/arm/st/stm32f1-pinctrl.dtsi:
```diff
@@ -0,0 +1,38 @@
+				};
+			};
+			usart2_pins_b: usart2@0 {
+				rx_tx {
```
Do you mean "usart2@1"?

erwango replied:
right, thanks

ydamigos commented on dts/arm/st/stm32l4.dtsi:
```diff
@@ -30,6 +32,11 @@
 
 	soc {
+		pinctrl: pin-controller {
+			compatible = "st,stm32-pinmux";
```
Should we add `#address-cells = <1>; #size-cells = <0>;`?

ydamigos commented on dts/arm/st/stm32f1.dtsi:
```diff
@@ -0,0 +1,89 @@
+
+	soc {
+		pinctrl: pin-controller {
+			compatible = "st,stm32-pinmux";
```
Should we add #address-cells = <1>; #size-cells = <0>;?

erwango replied:
I'm not sure why we should use 0 for size cells.
When we'll add gpio nodes, we'll have reg = <address size>;
Can you clarify?

ydamigos replied:
My suggestion was based on the usartx nodes which are added by this PR.
We could change to `#size-cells = <1>;` when we add the gpio nodes or now. Both are fine with me.

erwango replied:
ok, so done with  `#size-cells = <1>;`

ydamigos commented on dts/arm/st/stm32f3.dtsi:
```diff
@@ -0,0 +1,67 @@
+
+	soc {
+		pinctrl: pin-controller {
+			compatible = "st,stm32-pinmux";
```
Should we add #address-cells = <1>; #size-cells = <0>;?

erwango replied:
done with #size-cells = <1>;

ydamigos commented on dts/arm/st/stm32f4.dtsi:
```diff
@@ -30,6 +31,11 @@
 
 	soc {
+		pinctrl: pin-controller {
+			compatible = "st,stm32-pinmux";
```
Should we add #address-cells = <1>; #size-cells = <0>;?

ydamigos commented on dts/arm/st/stm32l4.dtsi:
```diff
@@ -0,0 +1,79 @@
+
+	soc {
+		pinctrl: pin-controller {
+			compatible = "st,stm32-pinmux";
```
Should we add #address-cells = <1>; #size-cells = <0>;?

erwango replied:
done with #size-cells = <1>;

erwango committed:
dts: arm: stm32l4: add pinctrl node and pins configurations

Add pinctrl node on stm32l4 soc dts files, including uart
pinctrl configurations. This is done thanks to <soc>-pinctrl.dtsi
file matching the <soc>.dtsi file
Populate stm32 l4 based boards dts files with references
to uart pinctrl nodes

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
dts: arm: stm32f4: Add pinctrl node and uart pin information

Add needed uart pinctrl configuration in pinmux node.
This is done thanks to <soc>-pinctrl.dtsi file matching
the <soc>.dtsi files
Populate stm32 f4 based boards dts files with references
to uart pinctrl nodes.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
dts: arm: stm32f3: Add pinctrl node and uart pin configurations

Add needed uart pinctrl configuration in pinmux node.
This is done thanks to <soc>-pinctrl.dtsi file matching
<soc>.dtsi files
Populate stm32 f3 based boards dts files with references
to uart pinctrl nodes

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
dts: arm: stm32f1: Add pinctrl node and uart pin information

Add needed uart pinctrl configuration in pinmux node.
Populate stm32 f1 based boards dts files with references
to uart pinctrl nodes

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
arch: stm32f1: rework pinmux driver

Rework stm32f1 pinmux code for future dts based pinmux code
generation.
Pin configuration is now done directly thanks to gpio port
configuration. Reference to pseudo alternate functions are
now removed same as the use of pins[] array.
Pins function (uart tx for instance) is set implicitly by
defining gpio mode and configuration.
This behavior is specific to stm32f10x series.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

galak commented on include/dt-bindings/pinctrl/stm32-pinctrl.h:
```diff
@@ -0,0 +1,335 @@
+#ifndef _STM32_PINCTRL_H_
+#define _STM32_PINCTRL_H_
+
+#ifdef CONFIG_SOC_SERIES_STM32F1X
```
rather than using CONFIG_* lets split this into different files:
stm32-f1x-pinctrl.h
stm32-f3-f4-l4-pinctrl.h

As mentioned before that the headers in here should be standalone from Zephyr

galak commented on include/dt-bindings/pinctrl/stm32-pinctrl.h:
```diff
@@ -0,0 +1,335 @@
+#define	STM32_PINMUX_FUNC_ALT_14 14
+#define	STM32_PINMUX_FUNC_ALT_15 15
+
+#define	ALT_FUNC_0 (STM32_PINMUX_FUNC_ALT_0 | STM32_MODER_ALT_MODE)
```
can we name space these like:

STM32_PINMUX_ALT_FUNC_{0..15}?

galak commented on include/dt-bindings/pinctrl/stm32-pinctrl.h:
```diff
@@ -0,0 +1,335 @@
+#define	STM32_PORTH 7	/* IO port H */
+
+
+/* override this at soc level */
```
Is this really needed in here?

galak commented:
Based on discussion with @erwango - tested on nucleo_f401, nucleo_f334r8, disco_l4, nucleo_f103

galak commented:
@ydamigos or @mbolivar can you try this out on a 96b_carbon?

ydamigos commented:
@galak I flashed the hello_world sample on 96b_carbon and it works fine. Should I try another sample?

mbolivar commented:
samples/basic/blinky and samples/subsys/shell are working fine on 96b_carbon with this.

erwango committed:
dts: arm: stm32: Add pinmux bindings and yaml

Add pinmux yaml file and bindings before introduction
of pinmux node in stm32 soc device tree files

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

ydamigos commented:
samples/subsys/shell is working fine also on olimexino_stm32 and stm32f3_disco

erwango committed:
arch: arm: stm32l4: Rework pinmux driver

STM32L4 pinmux handler is reworked to support future pinmux dts
generation.
Preliminary change is done to move pin configuration
informations in a {pin, conf} structure closer to dts fields
"pins" array is removed and information is transferred to
"pinconf" array

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers: pinmux: stm32: Clean up after rework

Following rework of pinmux driver for whole stm32 family,
remove call to stm32_get_pin_config which is no more
used.
Include dt-bindings stm32-pinctrl.h file to ensure coherency
between dts files and pinmux driver.
Due to change of "port" from enum to a series of define
(enum not accepted as dt binding), rework gpio_stm32_config
declaration.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
arch: arm: stm32f4: Rework pinmux handler

STM32F4 pinmux handler is reworked to support future pinmux dts
generation.
Preliminary change is done to move pin configuration
informations in a {pin, conf} structure closer to dts fields
"pins" array is removed as information is transfered to
"pinconf" array

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
arch: arm: stm32f3: Rework pinmux handler

STM32F3 pinmux handler is reworked to support future pinmux dts
generation.
Preliminary change is done to move pin configuration
informations in a {pin, conf} structure closer to dts fields
"pins" array is removed as information is transfered to
"pinconf" array

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
Updated according to comments.
Thanks @mbolivar, @ydamigos for testing!

galak commented on dts/arm/st/stm32f407.dtsi:
```diff
@@ -4,10 +4,15 @@
 	soc {
+		pinctrl: pin-controller {
+			reg = <0x40020000 0x2800>;
+		};
```
Should we have:

			compatible = "st,stm32-pinmux";
 +			#address-cells = <1>;
 +			#size-cells = <1>;
 +			reg = <0x40020000 0x1C00>;
 +

erwango replied:
Actually this is only the delta vs initial pincntrl node located in stm32f4.dtsi

mbolivar commented on include/dt-bindings/pinctrl/stm32-pinctrl.h:
```diff
@@ -0,0 +1,96 @@
+ * Applicable to STM32F3, STM32F4, STM32L4 series
+ */
+
+#define STM32_AFR_MASK			0xF
```
Shouldn't this be 0xFF?

erwango replied:
This is "Alternate function" [0,7]
This was already present and I removed the AFR_SHIFT in clean up since not used.
Maybe it could be made clearer.
Is this blocking? I'll have trouble to update this PR soon, but I would be pleased to make it clearer later on.

mbolivar replied:
If it's really 8 bits, isn't the mask 0xFF, not 0xF? Though the rest of the file makes it look like it's really just 4 bits, and the comment is wrong, should be:

```
 *    Alternate Function  [ 0 : 3 ]
 *    GPIO Mode           [ 4 : 5 ]
 *    GPIO Output type    [ 6 ]
 *    GPIO Speed          [ 7 : 8 ]
 *    GPIO PUPD config    [ 9 : 10 ]
```

erwango replied:
d'oh, you're right

mbolivar commented on arch/arm/soc/st_stm32/stm32l4/soc_gpio.c:
```diff
@@ -70,86 +49,28 @@ struct stm32l4x_gpio {
+
 		if (pud == GPIO_PUD_PULL_UP) {
-			*pincfg = STM32L4X_PIN_CONFIG_BIAS_PULL_UP;
+			*pincfg = *pincfg | STM32_PUPDR_PULL_UP;
```
What about output pull up or down? It seems like this only allows pull in the input direction.

erwango replied:
This was a pure conversion of the pre-existing code to the new format.

mbolivar replied:
Sure, makes sense
